clap_complete/src/dynamic/fish.rs

```diff
@@ -0,0 +1,130 @@
+//! Complete commands within fish
+
+// For fish the behaviour should match the default as close as possible
+// 1. grouping shorthand completions i.e. -a<TAB> should show other shorthands to end up with
+//    -ams...
+// 2. only complete options when one - is typed
+// Due to https://github.com/fish-shell/fish-shell/issues/7943 we need to implement this our self
+//
+
+use std::{
+    ffi::OsString,
+    fmt::Display,
+    io::{stdout, Write},
+    iter,
+};
+
+use clap::Args;
+use clap_lex::RawOsString;
+
+use super::{complete, completions_for_arg, Completer, CompletionContext};
+
+/// Dynamic completion for Fish
+pub struct Fish;
```

Comment:
Why is this commit in this PR?  Isn't #4177 responsible for all things fish

Reply:
I based this on the other branch, as I created the necessary API there.